lib/cartopy/geodesic/_geodesic.pyx

```diff
@@ -0,0 +1,253 @@
+                           flattening gives a prolate ellipsoid. If 
+                           flattening > 1, set flattening to 1/flattening.
+
+        """
+        # allocate some memory (filled with random data)
+        self.geod = <geod_geodesic*> PyMem_Malloc(sizeof(geod_geodesic))
+        if not self.geod:
+            raise MemoryError()
+        geod_init(self.geod, radius, flattening)
+
+    def direct(self, points, azimuths, distances):
+        """
+        Solve the direct geodesic problem where the length of the geodesic is 
+        specified in terms of distance.
+
+        Can accept and broadcast length 1 arguments. For example, given a single
+        start point and distance, an array of different azimuths can be supplied
+        to locate multiple endpoints.
+
+        Args:
+
+            * points - An n (or 1) by 2 numpy.ndarray, list or tuple of lon-lat
+                       points.
+                       The starting point(s) from which to travel.
+
+            * azimuths - A length n (or 1) numpy.ndarray or list of azimuth 
+                         values (degrees).
+
+            * distances - A length n (or 1) numpy.ndarray or list of distances 
+                          values (metres).
+
+        Returns:
+            An n by 3 np.ndarray of lons, lats, and azimuths of the located 
+            endpoint. 
+
+        """
+
+        cdef int n_points, i
+        cdef double[:, :] pts
+        cdef double[:] azims, dists
+
+        # Create numpy arrays from inputs, and ensure correct shape. Note: 
+        # reshape(-1) returns a 1D array from a 0 dimensional array as required 
+        # for broadcasting.
+        pts = np.array(points, dtype = np.float64).reshape((-1, 2))
+        azims = np.array(azimuths, dtype = np.float64).reshape(-1)
+        dists = np.array(distances, dtype = np.float64).reshape(-1)
+
+        n_points = max(pts.shape[0], azims.size, dists.size)
+
+        # Broadcast any length 1 arrays to the correct size.
+        try:
+            tmp = np.zeros((n_points, 2))
+            tmp[:, 0] += pts[:, 0]
+            tmp[:, 1] += pts[:, 1]
+
+            pts = tmp
+
+            azims = np.zeros(n_points) + azims[:]
+
+            dists = np.zeros(n_points) + dists[:]
+
+        except ValueError:
+            raise ValueError("Inputs must have common length n or length one.")
+
+        cdef double[:, :] return_pts = np.empty((n_points, 3))
+        cdef double[:] lon, lat, azi
+
+        lon = np.empty(n_points)
+        lat = np.empty(n_points)
+        azi = np.empty(n_points)
+
+        with nogil:
+            for i in prange(n_points):
+
+                geod_direct(self.geod, pts[i, 1], pts[i, 0], azims[i], dists[i], 
+                            &lat[i], &lon[i], &azi[i])
+                return_pts[i, 0] = lon[i]
+                return_pts[i, 1] = lat[i]
+                return_pts[i, 2] = azi[i]
+
+        return np.array(return_pts)
```

Comment:
I'm not completely familiar with Cython, so why isn't the result of `np.empty` not already a NumPy array?